File: 백준/Gold/1068. 트리/트리.js
const input = require('fs')
  .readFileSync('dev/stdin')
  .toString()
  .trim()
  .split('\n');

const N = Number(input.shift());
const deleteNode = Number(input.pop());
const nodeInfo = input[0].split(' ').map(Number);

const graph = Array.from({ length: N }, () => []);
let rootIdx = -1;
let count = 0;

nodeInfo.forEach((element, idx) => {
  if (element === -1) rootIdx = idx;
  else graph[element].push(idx);
});

function DFS(node) {
  for (let n of graph[node]) {
    DFS(n);
  }
  graph[node] = [];
  count += 1;
}

DFS(deleteNode);

let leafCount = 0;

for (let i = 0; i < N; i++) {
  if (graph[i].length === 0) leafCount += 1;
  else if (graph[i].length === 1 && graph[i][0] === deleteNode) leafCount += 1;
}
console.log(leafCount - count);

File: 백준/Gold/1068. 트리/트리.py
# 트리

import sys
sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline

N = int(input())
numList = list(map(int, input().split()))
deleteNode = int(input())
arr = [[] for _ in range(N)]
visited = [False] * N
rootNode = 0
leafCnt = 0

for i in range(N):
    if numList[i] == -1:
        rootNode = i
    else:
        arr[i].append(numList[i])
        arr[numList[i]].append(i)


def DFS(idx):
    global leafCnt
    visited[idx] = True
    isLeaf = True
    for v in arr[idx]:
        if not visited[v] and v != deleteNode:
            isLeaf = False
            DFS(v)
    if isLeaf:
        leafCnt += 1


if rootNode == deleteNode:
    print(0)
else:
    DFS(rootNode)
    print(leafCnt)
